### PR TITLE
[composite compliance] linalg_householder_product

### PR DESCRIPTION
**Pull request by kshitij12345**
Ref: #69991

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -5299,10 +5336,15 @@ Tensor householder_product_jvp(
 
     H_plus = apply_householder_reflector(v_i, sigma_i, H_plus, /*left=*/true);
 
-    dprod.add_(H_minus.matmul(
+    auto intermediate = H_minus.matmul(
```
Comment (kshitij12345) [inline torch/csrc/autograd/FunctionsManual.cpp:5339]: better name?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -5191,23 +5187,64 @@ std::tuple<Tensor, Tensor> householder_product_backward(
...
-    // K <- H_{i + 1}^{-1} @ K @ H_i
-    if (i < k - 1) {
-      auto v_i_next = input.narrow(-1, i + 1, 1);
-      auto s_i_next = sigma.narrow(-1, i + 1, 1);
-      K = apply_householder_reflector(
-          i + 1, v_i_next, s_i_next, K, /*left=*/true);
-      K = apply_householder_reflector(i, v_i, t_i, K, /*left=*/false);
+
+  Tensor input_grad, tau_grad;
+  if (areAnyTensorSubclassLike({input, tau, K})) {
```
Comment (kshitij12345) [inline torch/csrc/autograd/FunctionsManual.cpp:5192]: Add comment.

**Comment by kshitij12345** [comment]
@nikitaved (original author of the functions) can you please go through the changes :).

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -5191,23 +5187,69 @@ std::tuple<Tensor, Tensor> householder_product_backward(
...
+        input_grads.push_back(
+            at::zeros({}, input_.options()).expand_as(input_.select(-1, 0)));
+      }
+    }
+
+    input_grad = at::stack(input_grads, -1);
+    tau_grad = at::stack(tau_grads, -1);
+  } else {
+    input_grad = input_.new_zeros(input_.sizes());
+    tau_grad = tau.new_zeros(tau.sizes());
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:5234]: Question. Why `.new_zeros` and not `zeros_like`. Seems shorter as `sizes` do not have to be specified :)

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:5344]: Yes, let's call it `H_minus_dH_H_plus`.

Comment (kshitij12345) [reply torch/csrc/autograd/FunctionsManual.cpp:5344]: Sure. Thanks!

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 4]
```diff
@@ -5191,23 +5187,69 @@ std::tuple<Tensor, Tensor> householder_product_backward(
...
+      tau_grad.select(-1, i).copy_(tau_i_grad.squeeze(-1));
+
+      // K <- H_{i + 1}^{-1} @ K @ H_i
+      if (i < k - 1) {
+        auto v_i_next = input.narrow(-1, i + 1, 1);
+        auto s_i_next = sigma.narrow(-1, i + 1, 1);
+        K = apply_householder_reflector(
+            i + 1, v_i_next, s_i_next, K, /*left=*/true);
+        K = apply_householder_reflector(i, v_i, t_i, K, /*left=*/false);
+      }
```
[the remark this answers is not in the document]
Comment (kshitij12345) [reply torch/csrc/autograd/FunctionsManual.cpp:5252]: I agree. I thought/tried but since both have a slightly different structure, it seemed to me that duplicating made more sense than adding a bunch of branches.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 5]
```diff
@@ -5191,23 +5187,69 @@ std::tuple<Tensor, Tensor> householder_product_backward(
...
+    }
+
+    // zero gradients for the inputs not indexed by `k`.
+    if (k < input.size(-1)) {
+      for (const auto i : c10::irange(input.size(-1) - k)) {
+        std::ignore = i;
+        input_grads.push_back(
+            at::zeros({}, input_.options()).expand_as(input_.select(-1, 0)));
+      }
+    }
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:5228]: No need, you can just stack a block of zeros of size `(n, n - k)`.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 3 above)
Comment (kshitij12345) [reply torch/csrc/autograd/FunctionsManual.cpp]: Done.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 5 above)
Comment (kshitij12345) [reply torch/csrc/autograd/FunctionsManual.cpp]: Done. Thanks :)!

**Review by zou3519 — APPROVED** [review]

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 6]
```diff
@@ -5191,23 +5187,70 @@ std::tuple<Tensor, Tensor> householder_product_backward(
...
+
+    // Only first k columns are active in forward.
+    // zero gradients for the inactive input.
+    if (k < input.size(-1)) {
+      auto input_sizes = input_.sizes();
+      at::DimVector new_sizes(input_sizes);
+      new_sizes[input_.dim() - 1] = input.size(-1) - k;
+      auto zeros = at::zeros(new_sizes, input_.options());
+      input_grad = at::cat({input_grad, zeros}, -1);
+    }
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:5232]: I think we can do `stack` as the very last op, no? Zeros are better be pushed into the vector first, then we stack.

Comment (kshitij12345) [reply torch/csrc/autograd/FunctionsManual.cpp:5232]: `cat` behaves differently with vectors.

```python
>>> import torch
>>> x = torch.randn(5)
>>> torch.cat((x, x, x, x, x), -1).shape
torch.Size([25])
>>> torch.stack((x, x, x, x, x), -1).shape
torch.Size([5, 5])
```

Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:5232]: You just unsqueeze the last dim and that is it. Note that the code already does that.
